server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -725,10 +857,95 @@ private void handleFinalizationFailure(Exception e, SnapshotsInProgress.Entry en
                 new SnapshotException(snapshot, "Failed to update cluster state during snapshot finalization", e));
         } else {
             logger.warn(() -> new ParameterizedMessage("[{}] failed to finalize snapshot", snapshot), e);
-            removeSnapshotFromClusterState(snapshot, e);
+            removeSnapshotFromClusterState(snapshot, e,
+                    ActionListener.wrap(() -> runNextQueuedOperation(entry.repositoryStateId(), entry.repository())));
         }
     }
 
+    /**
+     * Run the next queued up repository operation for the given repository name.
+     * @param newGeneration current repository generation for given repository
+     * @param repository    repository name
+     */
+    private void runNextQueuedOperation(long newGeneration, String repository) {
+        synchronized (currentlyFinalizing) {
+            final Deque<SnapshotFinalization> outstandingForRepo = snapshotsToFinalize.get(repository);
+            final SnapshotFinalization nextFinalization;
+            if (outstandingForRepo == null) {
+                nextFinalization = null;
+            } else {
+                nextFinalization = outstandingForRepo.pollFirst();
+                if (outstandingForRepo.isEmpty()) {
+                    snapshotsToFinalize.remove(repository);
+                }
+            }
+            if (nextFinalization == null) {
+                final boolean removed = currentlyFinalizing.remove(repository);
+                assert removed;
+                runReadyDeletions();
+            } else {
+                logger.trace("Moving on to finalizing next snapshot [{}]", nextFinalization);
+                finalizeSnapshotEntry(nextFinalization.entry, nextFinalization.metadata, newGeneration);
+            }
+        }
+    }
+
+    /**
+     * Runs a cluster state update that checks whether we have any waiting snapshot deletions that can be executed and if so sets them
+     * to state {@link SnapshotDeletionsInProgress.State#META_DATA} and executes them.
+     *
+     * TODO: optimize this to execute in a single CS update together with finalizing the latest snapshot
```

Comment:
This is something to maybe handle before releasing (so that we never get a waiting delete that could be executed on master fail-over and can simplify the logic htere) but it requires non-trivial refactoring to make this happen and this PR is already huge and I figured it's better to get alignment on the strategy of how to merge this before doing that kind of refactoring.